clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProvider.java

```diff
@@ -120,36 +122,48 @@ private boolean refreshCredentials() throws IOException {
     final ZeebeClientCredentials fetchedCredentials = fetchCredentials();
     credentialsCache.put(endpoint, fetchedCredentials).writeCache();
 
-    if (fetchedCredentials.equals(credentials)) {
-      return false;
+    if (credentials == null || !credentials.isValid() || !fetchedCredentials.equals(credentials)) {
+      credentials = fetchedCredentials;
+      LOG.debug("Refreshed credentials.");
+
+      return true;
     }
 
-    credentials = fetchedCredentials;
-    LOG.debug("Refreshed credentials.");
-    return true;
+    return false;
   }
 
-  private static String createJsonPayload(final OAuthCredentialsProviderBuilder builder)
-      throws JsonProcessingException {
+  private static String createParams(final OAuthCredentialsProviderBuilder builder) {
     final Map<String, String> payload = new HashMap<>();
     payload.put("client_id", builder.getClientId());
     payload.put("client_secret", builder.getClientSecret());
     payload.put("audience", builder.getAudience());
     payload.put("grant_type", "client_credentials");
 
-    return JSON_MAPPER.writeValueAsString(payload);
+    final StringBuilder sb = new StringBuilder();
+    payload.forEach((key, value) -> sb.append(String.format("%s=%s&", encode(key), encode(value))));
```

Comment:
Nit: would this also work without having to deal with the awkward deletion of the last '&'?

```java
    payload.entrySet().stream()
        .map(e -> encode(e.getKey()) + "=" + encode(e.getValue()))
        .collect(Collectors.joining("&"));
```

Didn't test it, just off the top of my head, feel free to ignore if it seems more complicated - I just always find this extra deleting awkward

Reply:
I agree, that's a better solution